text/0004-new-pre-coordicide-consensus/0004-new-pre-coordicide-consensus.tex.md

```diff
@@ -0,0 +1,191 @@
+
+1. Look for milestone whose index is the `current latest solid milestone index + 1`. Milestones must be applied in order!
+1. When the milestone is found start traversing down the tangle in [DFS](https://en.wikipedia.org/wiki/Depth-first_search) going first to the trunk and only then the branch. Stop at the transactions that has only approved parents.
+2. Start applying bundles to the diffmap only as you go up the recursion stack (when you climb up back from the leafs).
+3. Every time you attempt to apply a bundle to the diffmap:
+   - If it is invalid mark it as `seen` and `ignored` by the milestone. 
+   - If it is valid and not conflicting with the current state then apply it to the ledger state. Mark it as `seen` and `approved`.
+   - If it is valid but conflicting with the current state then place it in an ordered *Conflict Set*.
+4. **Local overdraft handling**: After you are done traversing the subtangle attempt to apply again the conflicting bundles. Iterate the *Conflict Set* in order and attempt to apply the bundles to the ledger state. Mark the applied bundles as `seen` and `approved`. Remove applied bundles from the set. Repeat this step until the *Conflict Set* size doesn't change and mark the remaining bundles of the set as seen and ignored.
+
+**Note**
+Once a bundle is marked as ignored/seen/approved this will be final and it can't be changed by a later milestone that comes in.
+
+
+## Node Tip Selection
+
+Due to white-flag, no part of the tangle can be censored, thus the tangle can't be split by a double-spend. So as long as each new tip (bundle) a user creates approves two other random tips, the tangle shouldn't get divided into several subtangles. Thus we can get have a very fast tip selection by just selecting random tips.
+
+The problem is that lazy users may approve bundles or transactions that are not tips... This is a problem becuase lazy behavior can hinder confirmation rates. A lazy tip approves less bundles in the tangles. So we should have defenses in place to make sure that such lazy transactions will be left behind by honest tip-selection.
+
+So instead of total random tip selection we will do a weighted random tip selection. We will give each tip a score:
+
+```
+0 - If it is a lazy tip that shouldn't be selected
+1 - If it is a somewhat lazy tip
+2 - If it is a non-lazy good tip.
+```
+Since the actual consensus will be determined by milestone tip selection there can be more than one way to distribute the scores.
+We will propose two similar score systems that can be used by node implementations, a timestamp based one, and a milestone based one.
+
+But first some definitions:
+
+`Tip`:
+A solid bundle tail of a bundle that has no approvers.
+
+`Approved transaction Roots`:
+All `seened` (by milestone) transaction that can be reached by walking from a given transaction down to its parents. The walk must terminate once we reached a `seened` transaction.
```

Comment:
"down to its parents"? from context it seems like you mean traversing the past cone of a tx, but for clarity:
   What's a parent for transaction A? transaction B that is approved by A?

Reply:
yes, it is going to the past

I just made it clearer locally here